rocketpy/mathutils/function.py

```diff
@@ -333,142 +345,173 @@ def set_extrapolation(self, method="constant"):
-                    y = np.sum(coeffs * x ** np.arange(len(coeffs)))
-                else:
-                    # Extrapolate
-                    if extrapolation == 0:  # Extrapolation == zero
-                        y = 0
-                    elif extrapolation == 1:  # Extrapolation == natural
-                        y = 0
-                        for i, coef in enumerate(coeffs):
-                            y += coef * (x**i)
-                    else:  # Extrapolation is set to constant
-                        y = y_data[0] if x < x_min else y_data[-1]
-                return y
+                def natural_extrapolation(x, x_min, x_max, x_data, y_data, coeffs):
+                    if x < x_min:
+                        a = coeffs[:, 0]
+                        x = x - x_data[0]
+                    else:
+                        a = coeffs[:, -1]
+                        x = x - x_data[-2]
+                    y = a[3] * x**3 + a[2] * x**2 + a[1] * x + a[0]
+                    return y
 
-        elif self.__interpolation__ == "shepard":
-            # change the function's name to avoid mypy's error
-            def get_value_opt_multiple(*args):
-                return self.__interpolate_shepard__(args)
+            self._extrapolation_func = natural_extrapolation
+        elif extrapolation == 2:  # constant
 
-            get_value_opt = get_value_opt_multiple
+            def constant_extrapolation(x, x_min, x_max, x_data, y_data, coeffs):
+                return y_data[0] if x < x_min else y_data[-1]
 
-        try:
-            self.get_value_opt = get_value_opt
-        except UnboundLocalError:
-            warnings.warn(
-                "Cannot set the get_value_opt method when interpolation is "
-                f"{self.__interpolation}. Try using the set_interpolation method first."
-            )
+            self._extrapolation_func = constant_extrapolation
+
+    def set_get_value_opt(self):
+        """Defines a method that evaluates interpolations.
+
+        Returns
+        -------
+        self : Function
+        """
+        if callable(self.source):
+            self.get_value_opt = self.source
+        elif self.__dom_dim__ == 1:
+            self.get_value_opt = self._get_value_opt_1d
+        elif self.__dom_dim__ > 1:
+            self.get_value_opt = self._get_value_opt_nd
         return self
 
+    def _get_value_opt_1d(self, x):
+        """Evaluate the Function at a single point x. This method is used
+        when the Function is 1-D.
+
+        Parameters
+        ----------
+        x : scalar
+            Value where the Function is to be evaluated.
+
+        Returns
+        -------
+        y : scalar
+            Value of the Function at the specified point.
+        """
+        # Retrieve general info
+        x_data = self.x_array
+        y_data = self.y_array
+        x_min, x_max = self.x_initial, self.x_final
+        coeffs = self._coeffs
+        if x_min <= x <= x_max:
```

Comment:
Are you sure we should use `<=` in both the interval's extremes?
I was under impression at least the right side would receive a `<` instead.

I will have to think more about this point.

Reply:
It is either this or we could do:

```python
if x_min == x or x == x_max:
    return x
elif x_min < x < x_max:
    interpolate()
else:
    extrapolate()
```

Reply:
I agree with Stanos implementation, since the interpolation should keep all source points untouched, the x-extrema should be "interpolated".